vignettes/project2_chunks/project2_tim.Rmd

```diff
@@ -65,6 +56,512 @@ if (!require("tidySingleCellExperiment")) {
+  slice_sample %>%                    # "randomly slice out one row"
+  pull("cell") -> aCell               # "pull the column cell, assign to aCell"
+
+# your cell:
+show(aCell) 
+
+# does that mean we can ask for a random gene with certain attributes?
+as_tibble(rowData(tidybarnyard)) %>%  # "turn the rowData into a tibble"
+  filter(genome == "mm10") %>%        # "choose rows where genome == mm10"
+  slice_sample %>%                    # "randomly slice out one row"
+  pull("name") -> aGene               # "pull the column name, assign to aGene" 
+
+# your gene:
+show(aGene) 
+
+# how many copies of this random gene were found in this random cell? 
+counts(tidybarnyard)[aGene, aCell] 
+
+# note that the odds are good that you'll get a 0 for this random combination:
+library(Matrix)                                 # for the `nnzero` function 
+sparsity <- 1 - (nnzero(counts(tidybarnyard)) / # number of nonzero counts
+                 prod(dim(tidybarnyard)))       # number of rows * columns 
+
+# specifically, the chance of getting a zero is about 92.3%: 
+message(round(sparsity * 100, 1), "%")
+
+# This is fairly typical for a single-cell experiment, perhaps a bit low even.
+# You could also take 10000 or so random samples to estimate the sparsity.
+# For example, you could grab 100 genes from 100 cells at a time: 
+as_tibble(tidybarnyard) %>%       # turn tidybarnyard's colData into a tibble
+  slice_sample(n=100) %>%         # slice out 100 random rows, and then... 
+  pull("cell") -> aHundredCells   # assign the "cell" column to "aHundredCells"
+
+as_tibble(rowData(tidybarnyard)) %>% # turn tidybarnyard's rowData into a tibble
+  slice_sample(n=100) %>%            # slice out 100 random rows, and then 
+  pull("name") -> aHundredGenes      # assign "name" column to "aHundredGenes"
+
+samples <- (length(aHundredCells) * length(aHundredGenes))
+nonzero <- nnzero(counts(tidybarnyard)[aHundredGenes, aHundredCells])
+sparsity_hat <- (samples - nonzero) / samples 
+sparsity_hat # estimated sparsity
+
+# In fact, we can use this scheme to look at sampling error:
+sample_sparsity <- function(object, cells=100, genes=100) { 
+
+  samples <- cells * genes 
+  columns <- pull(slice_sample(as_tibble(object), n=cells), "cell")
+  rows <- pull(slice_sample(as_tibble(rowData(object)), n=genes), "name")
+  nonzero <- nnzero(counts(object)[rows, columns])
+  sparsity <- (samples - nonzero) / samples
+  return(sparsity) 
+
+}
+
+# the `replicate` function allows us to apply this many times over:
+estimates <- replicate(n=100, sample_sparsity(tidybarnyard)) # "do it 100 times"
+
+# plot estimates: 
+library(ggplot2)
+ggplot(tibble(estimate=estimates), aes(estimate)) + 
+  geom_histogram() + 
+  geom_vline(xintercept=sparsity, color="red", lwd=3) +
+  theme_minimal() + 
+  ggtitle("Sparsity of UMI matrix (true value in red)")
+
+# the Central Limit Theorem lives to fight another day,
+# and we have a decent idea of how to navigate our data.
```

Comment:
What is this section useful for? Why do we need to see if our data has a normal distribution? How does that inform how we navigate our data?

Reply:
see expanded resampling and direct confidence interval calculation section.  leads into project 3, how many observations do you need to exclude the possibility of (whatever) at (some alpha level) for (some % of the time).